### PR TITLE
pkg: add FlashDB

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This adds a package for [FlashDB](http://armink.gitee.io/flashdb/), an embedded time series and key-value data base.

It can either operate directly on a raw MTD device (`flashdb_mtd`) or use the VFS layer for file based operation.

### Testing procedure

I added a test that uses the examples provided by FlashDB.

<details><summary>tests/pkg_flashdb</summary>

```
2022-02-05 00:21:09,418 # main(): This is RIOT! (Version: 2022.04-devel-117-gda35d-pkg/flashdb)
2022-02-05 00:21:09,591 # ==================== kvdb_basic_sample ====================
2022-02-05 00:21:09,601 # get the 'boot_count' value is 1
2022-02-05 00:21:11,210 # set the 'boot_count' value to 2
2022-02-05 00:21:11,215 # ===========================================================
2022-02-05 00:21:11,220 # ================= kvdb_type_string_sample =================
2022-02-05 00:21:12,636 # create the 'temp' string KV, value is: 36C
2022-02-05 00:21:12,642 # get the 'temp' value is: 36C
2022-02-05 00:21:14,057 # set 'temp' value to 38C
2022-02-05 00:21:14,340 # delete the 'temp' finish
2022-02-05 00:21:14,346 # ===========================================================
2022-02-05 00:21:14,351 # ================== kvdb_type_blob_sample ==================
2022-02-05 00:21:15,241 # create the 'temp' blob KV, value is: 36
2022-02-05 00:21:15,272 # get the 'temp' value is: 36
2022-02-05 00:21:16,689 # set 'temp' value to 38
2022-02-05 00:21:16,972 # delete the 'temp' finish
2022-02-05 00:21:16,977 # ===========================================================
2022-02-05 00:21:17,013 # ======================= tsdb_sample =======================
2022-02-05 00:21:17,017 # append the new status.temp (36) and status.humi (85)
2022-02-05 00:21:17,022 # append the new status.temp (38) and status.humi (90)
2022-02-05 00:21:17,028 # [query_cb] queried a TSL: time: 19, temp: 36, humi: 85
2022-02-05 00:21:17,034 # [query_cb] queried a TSL: time: 21, temp: 38, humi: 90
2022-02-05 00:21:17,060 # [query_by_time_cb] queried a TSL: time: 19, temp: 36, humi: 85
2022-02-05 00:21:17,066 # [query_by_time_cb] queried a TSL: time: 21, temp: 38, humi: 90
2022-02-05 00:21:17,084 # query count is: 0
2022-02-05 00:21:17,093 # set the TSL (time 19) status from 3 to 3
2022-02-05 00:21:17,374 # set the TSL (time 21) status from 3 to 3
2022-02-05 00:21:17,673 # ===========================================================
```
</details>

<details><summary>tests/pkg_flashdb_mtd</summary>

```
2022-02-05 00:12:48,097 # main(): This is RIOT! (Version: 2022.04-devel-118-g80d9a-pkg/flashdb)
2022-02-05 00:12:48,104 # [I/FAL] Flash Abstraction Layer (V0.5.0) initialize success.
2022-02-05 00:12:48,117 # ==================== kvdb_basic_sample ====================
2022-02-05 00:12:48,120 # get the 'boot_count' value is 12
2022-02-05 00:12:48,132 # set the 'boot_count' value to 13
2022-02-05 00:12:48,137 # ===========================================================
2022-02-05 00:12:48,143 # ================= kvdb_type_string_sample =================
2022-02-05 00:12:48,152 # create the 'temp' string KV, value is: 36C
2022-02-05 00:12:48,155 # get the 'temp' value is: 36C
2022-02-05 00:12:48,157 # set 'temp' value to 38C
2022-02-05 00:12:48,160 # delete the 'temp' finish
2022-02-05 00:12:48,165 # ===========================================================
2022-02-05 00:12:48,170 # ================== kvdb_type_blob_sample ==================
2022-02-05 00:12:48,180 # create the 'temp' blob KV, value is: 36
2022-02-05 00:12:48,182 # get the 'temp' value is: 36
2022-02-05 00:12:48,185 # set 'temp' value to 38
2022-02-05 00:12:48,187 # delete the 'temp' finish
2022-02-05 00:12:48,192 # ===========================================================
2022-02-05 00:12:48,198 # ======================= tsdb_sample =======================
2022-02-05 00:12:48,203 # append the new status.temp (36) and status.humi (85)
2022-02-05 00:12:48,207 # append the new status.temp (38) and status.humi (90)
2022-02-05 00:12:48,212 # [query_cb] queried a TSL: time: 1, temp: 36, humi: 85
2022-02-05 00:12:48,217 # [query_cb] queried a TSL: time: 1, temp: 38, humi: 90
2022-02-05 00:12:48,222 # [query_by_time_cb] queried a TSL: time: 1, temp: 36, humi: 85
2022-02-05 00:12:48,228 # [query_by_time_cb] queried a TSL: time: 1, temp: 38, humi: 90
2022-02-05 00:12:48,229 # query count is: 0
2022-02-05 00:12:48,233 # set the TSL (time 1) status from 3 to 3
2022-02-05 00:12:48,236 # set the TSL (time 1) status from 3 to 3
2022-02-05 00:12:48,242 # ===========================================================
```
</details>

### Issues/PRs references

includes #17341 for convenience 
[repo of FlashDB](https://github.com/armink/FlashDB)

### CODE REVIEW

**Comment by maribu** [comment]
I like this one :) The CI has some comments.

**Comment by fjmolinas** [comment]
One question out of curiosity, the GitHub page sys its supports wear balance, do you know how that is done? I was trying to find some information to estimate flash wear out when using the API but didn't find a section on that immediately.

**Comment by fjmolinas** [comment]
I think python needs a rebase to be fixed?

File: tests/pkg_flashdb/main.c
Comment (kfessel) [inline tests/pkg_flashdb/main.c:65]: this line requires USE_MODULE += libc_gettimeofday

**Comment by benpicco** [comment]
I added a `flashdb_vfs` module to make the naming more consistent. Now `flashdb_mtd` selects the MTD backend and `flashdb_vfs` the file backend.

Also no point in keeping the MTD commit separate.

File: pkg/flashdb/mtd/fal_mtd_port.c
Comment (benpicco) [reply pkg/flashdb/mtd/fal_mtd_port.c:68]: Hm should we just blacklist `periph_flashpage`/add a warning for now or wait for #17683?

Comment (fjmolinas) [reply pkg/flashdb/mtd/fal_mtd_port.c:68]: ping @chrysn!

File: Makefile.dep
Comment (fjmolinas) [inline Makefile.dep:84]: Unrelated

File: pkg/flashdb/Kconfig
Comment (fjmolinas) [inline pkg/flashdb/Kconfig:10]: Missing a lot modules here ;)

File: pkg/flashdb/Makefile.include
Comment (fjmolinas) [inline pkg/flashdb/Makefile.include:8]: Could you document these pseudomodules?

File: pkg/flashdb/include/fal_cfg.h
Comment (fjmolinas) [inline pkg/flashdb/include/fal_cfg.h:50]: Can you brake this line?

File: tests/pkg_flashdb_mtd/Makefile
Comment (fjmolinas) [inline tests/pkg_flashdb_mtd/Makefile:19]: Any reasoning for these lengths?

**Review by fjmolinas — COMMENTED** [review]
Some comments

File: Makefile.dep
Comment (benpicco) [reply Makefile.dep:84]: Doesn't it make sense to fix the whitespace when touching that file?

File: pkg/flashdb/Makefile.include
Comment (benpicco) [reply pkg/flashdb/Makefile.include:8]: Is Kconfig enough for this?

File: pkg/flashdb/Makefile.dep
Comment (chrysn) [inline pkg/flashdb/Makefile.dep:19]: Doesn't this also enable FAL? (It does in Kconfig AIU).

File: pkg/flashdb/include/fal_cfg.h
Comment (chrysn) [inline pkg/flashdb/include/fal_cfg.h:36]: flashDB boasts that it can do multiple databases on the same system; can we?

It's OK if there is a choice that we'll have a single system-wide KVDB in place, but then the rationale for that should be outlined, and maybe the database initialization should be outside the user application (because then the OS provides "the single" KVDB as a readily usable service).

File: pkg/flashdb/include/fal_cfg.h
Comment (chrysn) [inline pkg/flashdb/include/fal_cfg.h:49]: Given that there is only one MTD device, is this a dummy partition table, and if so, why 4 partitions?

File: pkg/flashdb/Kconfig
Comment (benpicco) [reply pkg/flashdb/Kconfig:26]: A user shouldn't be able to toggle this - this is only needed for MTD access and is a dependency of `flashdb_mtd`

Comment (chrysn) [reply pkg/flashdb/Kconfig:26]: If VFS doesn't use this, why is this a separate flag from FLASHDB_MTD at all?

File: pkg/flashdb/include/fal_cfg.h
Comment (benpicco) [reply pkg/flashdb/include/fal_cfg.h:36]: Yea that's an unfortunate consequence of the MTD <-> FAL adaption layer. FAL only expects a single device of an implementation to exist, there is only `write(offset, data, len)` with no way to specify a context pointer to discriminate between different devices. (https://github.com/armink/FlashDB/issues/132#issuecomment-1033138711)

A possible solution would be mapping different MTD backends to different virtual addresses, but I didn't want to go down that rabbit hole.

File: pkg/flashdb/include/fal_cfg.h
Comment (chrysn) [reply pkg/flashdb/include/fal_cfg.h:49]: But we have only one partition by the above, and a brief search in upstream sources shows that there don't have to be 4 entries.

Comment (benpicco) [reply pkg/flashdb/include/fal_cfg.h:49]: A user can specify up to 4 partitions before having to touch this file :wink:

Comment (chrysn) [reply pkg/flashdb/include/fal_cfg.h:49]: I thought that by https://github.com/RIOT-OS/RIOT/pull/17612/files#r897214173 we can only have one backend ... maybe it'd help adding a few words on what partitions mean here.

File: pkg/flashdb/Kconfig
Comment (benpicco) [reply pkg/flashdb/Kconfig:26]: It's a separate module in FlashDB, so it felt most natural to also model it as a separate module in RIOT

`flashdb_fal` lives in `$(PKG_SOURCE_DIR)/port/fal/src`
`flashdb_mtd` lives in `pkg/flashdb/mtd`

File: pkg/flashdb/include/fal_cfg.h
Comment (benpicco) [reply pkg/flashdb/include/fal_cfg.h:49]: One backend can have multiple partitions.

Comment (chrysn) [reply pkg/flashdb/include/fal_cfg.h:49]: So multiple partitions are usable, but because FlashDB expects them all to be provided on a single continuous address (rather than in a (partition, address) addressing), any and all must be on the same MTD device. Got it now, but it'd be good to have this information accessible to someone using the module after it's merged, along with a pointer to where the choice of MTD partition goes when initializing the database.

Comment (benpicco) [reply pkg/flashdb/include/fal_cfg.h:55]: Like this? 7db5c1e084be3c4070bd3a7dde7cb96803817fde

**Comment by benpicco** [comment]
Ping?

File: tests/pkg_flashdb_mtd/Makefile
Comment (aabadie) [reply tests/pkg_flashdb_mtd/Makefile:19]: Then is it possible to make these values configurable somewhere ? Maybe in the package Makefile.include or something like this that works.

Comment (aabadie) [reply tests/pkg_flashdb_mtd/Makefile:19]: Sorry, I didn't notice those were set in the application Makefile. So good for me like this :)

File: pkg/flashdb/Makefile
Comment (kaspar030) [inline pkg/flashdb/Makefile:3]:
```suggestion
# 1.2.0
PKG_VERSION=03500fa92568a0fba944ed44877fb72ec74fd90c
```

**Review by kaspar030 — CHANGES_REQUESTED** [review]

File: pkg/flashdb/Makefile.include
Comment (kaspar030) [reply pkg/flashdb/Makefile.include:8]: to me it is.

**Comment by kaspar030** [comment]
Code / integration looks fine to me. Testing delegated to the included test applications.
Didn't look at the Kconfig. @chrysn please come take another look!

**Review by MrKevinWeiss — COMMENTED** [review]
Just a note from the kconfig side:

- This is not actually using kconfig dep resolution since there is not `app.test.config` in the test folder
- The  use of `KCONFIG_USEPKG_FLASHDB` means that kconfig is actually getting information from make dependency resolution and I guess is only used for configuring values.
- In order to inject specific defines for package purposes (since Kconfig symbols resolve to `CONFIG_<symbol>` you need a translation step in the package such as https://github.com/RIOT-OS/RIOT/blob/master/pkg/tinydtls/Makefile.include#L53

I wish I could take the time to provide a simple diff of the TEST_KCONFIG=1 resolution but I am still working on a USB abstraction.

I don't want the TEST_KCONFIG stuff blocking though so I would expect that, since there is not `app.config.test` it won't actually fail.  You can then iterate, use, and improve until I or someone has some time to implement the dependencies in Kconfig.

BTW, will it always be the user who wants to select which backend to use or will we eventually have some "smart selection" where we would say, I want `flashdb` and the system know what backend is best?

**Comment by MrKevinWeiss** [comment]
Also is it correct that `flashdb_vfs` and `flashdb_mtd` are mutually exclusive?

**Comment by benpicco** [comment]
> will it always be the user who wants to select which backend

yes. even if you have a file system on e.g. the SD card, you still might want to use FlashDB on the internal flash instead - or not

> Also is it correct that flashdb_vfs and flashdb_mtd are mutually exclusive?


Yes

**Comment by maribu** [comment]
@kaspar030 or @chrysn since you started the review, do you have time for this? Otherwise I could take over.

**Comment by benpicco** [comment]
> do you have time for this?

I guess that's a no

**Comment by chrysn** [comment]
I'm very sorry: my time is terribly hard to predict right now, and I'm
subscribed to too many issues (maybe I can make posts-that-ping-me
highlighted among the others?). I'd appreciate if someone else could
continue reviewing this.

File: pkg/flashdb/doc.txt
Comment (kfessel) [inline pkg/flashdb/doc.txt:2]:
```suggestion
 * @defgroup pkg_flashdb    flashdb_*: Database for Flash 
```

I'd like to see the group heading like this. or without the *

**Review by kfessel — COMMENTED** [review]
There is no reason to further hold this back (review stages 1 to 5  are already green by kaspar) jue did a hardware test i did run the test on native and read this.

Issues raised earlier are addressed.

Please squash and activate CI.


I would like to see the pseudomodules listed in Doxygen.
I think there should be some kind of warning that this will not do wear leveling and advising a solution like littlefs2.
I also would like to see the title of the Doxygen group changed a little for more userfullnes of that.

Please squash these changes right in.

**Comment by benpicco** [comment]
> I think there should be some kind of warning that this will not do wear leveling and advising a solution like littlefs2.

Whaat? I thought that was one of it's selling points.

**Comment by kfessel** [comment]
> > I think there should be some kind of warning that this will not do wear leveling and advising a solution like littlefs2.
> 
> Whaat? I thought that was one of it's [selling points](http://armink.gitee.io/flashdb/#/README?id=key-features).

maybe i missed something or the wearleveling is done another way than in littlefs2.
The memory-file for (mtd) on native has one instance of the datebase after multiple runs.
The littlefs2 memoryfile (vfs version) has the database multiple times (every time written to another place)
Maybe this does not level wear across reboot?

**Comment by kfessel** [comment]
seems like the compactness of the written data (littlefs2 is more spacey when writing to different places) in the mtd threw me off -> it writes to different places (levels the wear)

File: pkg/flashdb/doc.txt
Comment (kfessel) [inline pkg/flashdb/doc.txt:2]: might be better to go with 

```suggestion
 * @defgroup pkg_flashdb   flashdb: Database for Flash
```

but this is better then most modules so this comment might as well be ignored.

https://github.com/RIOT-OS/RIOT/issues/7094 < this is an age old RFC that is discussing the these goup titles

Comment (benpicco) [reply pkg/flashdb/doc.txt:2]: But that’s the name of the software - and the needed module names are explained just below that.

**Comment by kfessel** [comment]
I had another look at the writes (native mtd debug)-> only writes are happening  no erases (i assume until rollover) 

There is a huge number of small writes  (sometimes to the same location but no erase) the test has a increasing number of writes with every run as the database increases with each run by some tvdb entries. I think this is ok.

<details>

```
grep write 
mtd_native: write from page 4, offset 0xa3 count 1
mtd_native: write from page 5, offset 0x47 count 1
mtd_native: write from page 5, offset 0x4b count 20
mtd_native: write from page 5, offset 0x5f count 10
mtd_native: write from page 5, offset 0x69 count 4
mtd_native: write from page 5, offset 0x47 count 1
mtd_native: write from page 4, offset 0xa3 count 1
mtd_native: write from page 5, offset 0x6d count 1
mtd_native: write from page 5, offset 0x71 count 20
mtd_native: write from page 5, offset 0x85 count 4
mtd_native: write from page 5, offset 0x89 count 3
mtd_native: write from page 5, offset 0x6d count 1
mtd_native: write from page 5, offset 0x6d count 1
mtd_native: write from page 5, offset 0x8c count 1
mtd_native: write from page 5, offset 0x90 count 20
mtd_native: write from page 5, offset 0xa4 count 4
mtd_native: write from page 5, offset 0xa8 count 3
mtd_native: write from page 5, offset 0x8c count 1
mtd_native: write from page 5, offset 0x6d count 1
mtd_native: write from page 5, offset 0x8c count 1
mtd_native: write from page 5, offset 0xab count 1
mtd_native: write from page 5, offset 0xaf count 20
mtd_native: write from page 5, offset 0xc3 count 4
mtd_native: write from page 5, offset 0xc7 count 4
mtd_native: write from page 5, offset 0xab count 1
mtd_native: write from page 5, offset 0xab count 1
mtd_native: write from page 5, offset 0xcb count 1
mtd_native: write from page 5, offset 0xcf count 20
mtd_native: write from page 5, offset 0xe3 count 4
mtd_native: write from page 5, offset 0xe7 count 4
mtd_native: write from page 5, offset 0xcb count 1
mtd_native: write from page 5, offset 0xab count 1
mtd_native: write from page 5, offset 0xcb count 1
mtd_native: write from page 21, offset 0x8 count 1
mtd_native: write from page 21, offset 0xc count 12
mtd_native: write from page 2f, offset 0x88 count 8
mtd_native: write from page 21, offset 0x8 count 1
mtd_native: write from page 21, offset 0x18 count 1
mtd_native: write from page 21, offset 0x1c count 12
mtd_native: write from page 2f, offset 0x80 count 8
mtd_native: write from page 21, offset 0x18 count 1
mtd_native: write from page 20, offset 0x28 count 1
mtd_native: write from page 20, offset 0x38 count 1
mtd_native: write from page 20, offset 0x48 count 1
mtd_native: write from page 20, offset 0x58 count 1
mtd_native: write from page 20, offset 0x68 count 1
mtd_native: write from page 20, offset 0x78 count 1
mtd_native: write from page 20, offset 0x88 count 1
mtd_native: write from page 20, offset 0x98 count 1
mtd_native: write from page 20, offset 0xa8 count 1
mtd_native: write from page 20, offset 0xb8 count 1
mtd_native: write from page 20, offset 0xc8 count 1
mtd_native: write from page 20, offset 0xd8 count 1
mtd_native: write from page 20, offset 0xe8 count 1
mtd_native: write from page 20, offset 0xf8 count 1
mtd_native: write from page 21, offset 0x8 count 1
mtd_native: write from page 21, offset 0x18 count 1
```

</details>

File: pkg/flashdb/mtd/fal_mtd_port.c
Comment (kfessel) [inline pkg/flashdb/mtd/fal_mtd_port.c:62]: The documentation says 

```
/* the flash write granularity, unit: bit
- * only support 1(nor flash)/ 8(stm32f2/f4)/ 32(stm32f1) */
```

Should we have an assert to ensure that we are in spec of flashdb?

and isn't ` _mtd->write_size  ^= to write_gran=8`
```suggestion
    mtd_flash0.write_gran   = _mtd->write_size == 1 ?
                            8 :
                            _mtd->write_size == 4 ? 
                            32 : 
                            (assert( _mtd->write_size == 1||_mtd->write_size == 4),0);
```

Comment (kfessel) [reply pkg/flashdb/mtd/fal_mtd_port.c:62]: according to the flashdb documentation the stm32f2 requires `write gran = 8` but we can not set this atm s`ince _mtd->write_size == 1`  would be `write gran 8` but is set to 1

Comment (benpicco) [reply pkg/flashdb/mtd/fal_mtd_port.c:62]: Yea that's a bit odd, I don't quite understand how write granularity of 1 bit works.
You don't even have 1 bit writes in RAM unless you do bit-banding.

Comment (kfessel) [reply pkg/flashdb/mtd/fal_mtd_port.c:62]: i think its not about the write size but the the size of the modification erases flash byte starts at `b11111111` with `write gran 1` next state might be  `b1111 1110` and the next  `b1111 1100` but with write gran 8 the second step is not allowed  therefor the db has to use the next byte.

Comment (kfessel) [reply pkg/flashdb/mtd/fal_mtd_port.c:62]: i think its not about the write size but the the size of the modification erases flash byte starts at `b11111111` with `write gran 1` next state might be  `b1111 1110` and the next  `b1111 1100` but with write gran 8 the second step is not allowed  therefor the db has to use the next byte.

<RIOT>/build/pkg/flashdb/src/fdb_utils.c :L95 has a table that might explain that

I am not sure how to deal with that `write gran 1` has wear leveling advantages but we don't know weather the mtd chosen has byte granularity for its modifies or bit (in both cases we would have write_size 1)
This i especially tricky  since one of the examples given for 8 are the stm32f2/f4  and we got a good amount of the nucleos in our boards fall in that range.

8 would be a safe fallback. (the flashdb falls back to higher if two flash are used with different granularity)

i also saw that they themself make use of other granularities that 1/8/32 ->


```suggestion
    mtd_flash0.write_gran   = _mtd->write_size * 8;
```

and maybe some mechanism to overwrite that

Comment (benpicco) [reply pkg/flashdb/mtd/fal_mtd_port.c:62]: Looking though the code of FlashDB I can't even find a place where `write_gran` is actually used outside a demo :exploding_head:

Comment (kfessel) [reply pkg/flashdb/mtd/fal_mtd_port.c]: Seems a bit like the struct member was a red herring the code for the stauts writes makes use of the define. 

[e.g.: here](https://github.com/armink/FlashDB/blob/master/src/fdb_utils.c#L155)

Is that define `FDB_WRITE_GRAN` set by this pkg ?

Comment (benpicco) [reply pkg/flashdb/mtd/fal_mtd_port.c]: Hm I overlooked that too. How would you suggest to set this? It has to be set at compile time - right now it's just always the default value unless set by the user.

Comment (benpicco) [reply pkg/flashdb/mtd/fal_mtd_port.c]: Ok, I've set it to 1, `FDB_WRITE_GRAN` is already 1 by default.

Comment (kfessel) [reply pkg/flashdb/mtd/fal_mtd_port.c]: To double check that this is unused (write only value), I removed front the struct an init() and I got no compiler complains

**Review by kfessel — APPROVED** [review]
Squash and Go

**Comment by benpicco** [comment]
Thank you for the review!

bors merge

**Comment by benpicco** [comment]
bors merge

**Comment by benpicco** [comment]
bors merge

**Pull request merged**
